QUIC wire protocol replacement

tewinget opened the pull request.

tewinget committed:
Add oxen-libquic submodule

TODO: correctly plug it into CMake and remove ngtcp2 submodule from lokinet
as it will now come from oxen-libquic, if even used directly by lokinet at all.

tewinget committed:
initial quic wire proto Connection skeleton

tewinget committed:
initial quic wire proto Endpoint skeleton

dr7ana committed:
cmake tweaks for redundant dependencies (WIP)

majestrate commented on llarp/link/connection.hpp:
```diff
@@ -0,0 +1,30 @@
+    public:
+
+      const RouterContact& RemoteRC() { return remote_rc; }
+      const RouterID& RemoteID() { return remote_id; }
```
will this `RouterID` not just be the identity pubkey from the router contact?

dr7ana replied:
resolving these on the WIP pr, messaged you offline

majestrate commented on llarp/link/connection.hpp:
```diff
@@ -0,0 +1,30 @@
+
+    public:
+
+      const RouterContact& RemoteRC() { return remote_rc; }
```
lacks const correctness.

majestrate commented on llarp/link/connection.hpp:
```diff
@@ -0,0 +1,30 @@
+
+    RouterID remote_id;
+    RouterContact remote_rc;
+    AddressInfo remote_addr_info; // RC may have many, this is the one in use for this connection
```
could the `remote_addr_info` member be a const l value reference instead?

majestrate commented on llarp/quic/endpoint.cpp:
```diff
@@ -130,7 +130,7 @@ namespace llarp::quic
   void
   Endpoint::handle_conn_packet(Connection& conn, const Packet& p)
   {
-    if (ngtcp2_conn_is_in_closing_period(conn))
```
should this be cherry-picked into its own commit?

tewinget committed:
Partial implementation of libquic as wire protocol

TODO:

- set up all the callbacks for libquic

- define control message requests, responses, commands

- plug new control messages into lokinet (path creation, network state, etc)

- plug connection state changes (established, failed, closed, etc.) into lokinet

- lots of cleanup and miscellanea
